pkg/csi/service/fcd/controller.go

```diff
@@ -72,15 +74,112 @@ func (c *controller) CreateVolume(
 	req *csi.CreateVolumeRequest) (
 	*csi.CreateVolumeResponse, error) {
 
-	return nil, nil
+	params := req.GetParameters()
+
+	// Volume Name
+	volName := req.GetName()
+	if len(volName) == 0 {
+		myUUID, err := uuid.NewUUID()
+		if err != nil {
+			log.Errorf("Failed to generate a UUID. Err: %v", err)
+			return nil, err
+		}
+		volName = myUUID.String()
+	}
+
+	// Volume Size - Default is 10 GiB
+	volSizeBytes := int64(DefaultGbDiskSize * GbInBytes)
+	if req.GetCapacityRange() != nil {
+		volSizeBytes = int64(req.GetCapacityRange().GetRequiredBytes())
+	}
+	volSizeMB := int64(volumeutil.RoundUpSize(volSizeBytes, GbInBytes)) * 1024
+
+	// Volume Type
+	datastoreType := vclib.TypeDatastoreCluster
+	volType := params[AttributeFirstClassDiskParentType]
+	if volType == string(vclib.TypeDatastore) {
+		datastoreType = vclib.TypeDatastore
+	}
+
+	datastoreName := params[AttributeFirstClassDiskParentName]
+
+	//TODO: zones are currently unimplemented. supports single VC/DC only.
+	// Please see function for more details
+	zone := "TODO"
+	discoveryInfo, err := c.connMgr.WhichVCandDCByZone(ctx, zone)
+	if err != nil {
+		log.Errorf("Failed to retrieve VC/DC based on zone %s. Err: %v", zone, err)
+		return nil, err
+	}
+
+	err = discoveryInfo.DataCenter.CreateFirstClassDisk(ctx, datastoreName, datastoreType, volName, volSizeMB)
+	if err != nil {
+		log.Errorf("CreateFirstClassDisk failed. Err: %v", err)
+		return nil, err
+	}
+
+	firstClassDisk, err := discoveryInfo.DataCenter.GetFirstClassDiskByName(ctx, datastoreName, datastoreType, volName)
+	if err != nil {
+		log.Errorf("GetFirstClassDiskByName failed. Err: %v", err)
+		return nil, err
+	}
+
+	attributes := make(map[string]string)
+	attributes[AttributeFirstClassDiskType] = FirstClassDiskTypeString
+	attributes[AttributeFirstClassDiskName] = firstClassDisk.Config.Name
+	attributes[AttributeFirstClassDiskParentType] = string(firstClassDisk.ParentType)
+	if firstClassDisk.ParentType == vclib.TypeDatastoreCluster {
+		attributes[AttributeFirstClassDiskParentName] = firstClassDisk.StoragePodInfo.Summary.Name
+		attributes[AttributeFirstClassDiskOwningDatastore] = firstClassDisk.DatastoreInfo.Info.Name
+	} else {
+		attributes[AttributeFirstClassDiskParentName] = firstClassDisk.DatastoreInfo.Info.Name
+	}
+
+	resp := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			Id:            firstClassDisk.Config.Id.Id,
+			CapacityBytes: int64(units.FileSize(firstClassDisk.Config.CapacityInMB * MbInBytes)),
+			Attributes:    attributes,
+			//TODO: ContentSource?
+			//TODO: AccessibleTopology, needed when we support multiple VC/DC combos
+		},
+	}
+
+	return resp, nil
 }
 
 func (c *controller) DeleteVolume(
 	ctx context.Context,
 	req *csi.DeleteVolumeRequest) (
 	*csi.DeleteVolumeResponse, error) {
 
-	return nil, nil
+	params := req.GetControllerDeleteSecrets()
+
+	// Volume Type
+	datastoreType := vclib.TypeDatastoreCluster
+	volType := params[AttributeFirstClassDiskParentType]
```

Comment:
same as before, I wonder what happens when this parameter is not present.

Furthermore, in 1.0 spec, there are only two fields. volume ID, and secrets. Secrets are meant to be keys, not the volume context/parameters from before. So that brings up an interesting question of how we are going to get that information, because it does not look available to me.

Reply:
Will cross that road when we update to 1.0

Reply:
Will be making changes based on the new 1.0 spec